pkg/router/controller/status.go

```diff
@@ -133,40 +168,95 @@ func performIngressConditionUpdate(action string, lease writerlease.Lease, track
+func performIngressConditionRemoval(action string, lease writerlease.Lease, tracker ContentionTracker, oc client.RoutesGetter, lister routelisters.RouteLister, route *routev1.Route, routerName string, condType routev1.RouteIngressConditionType) {
+	// Key the lease's work off of the route UID and the condition type, as different conditions will require separate updates.
+	workKey := createWorkKey(route.UID, condType)
+	routeNamespace, routeName := route.Namespace, route.Name
+	oldRouteUID := route.UID
+
+	lease.Try(workKey, func() (writerlease.WorkResult, bool) {
+		route, err := lister.Routes(routeNamespace).Get(routeName)
+		if err != nil {
+			return writerlease.None, false
+		}
+		if route.UID != oldRouteUID {
+			log.V(4).Info("skipped update due to route UID changing (likely delete and recreate)", "action", action, "namespace", route.Namespace, "name", route.Name)
+			return writerlease.None, false
+		}
+
+		route = route.DeepCopy()
+		changed, now, latest, original := removeIngressCondition(route, routerName, condType)
+		if !changed {
+			log.V(4).Info("no changes to route needed", "action", action, "namespace", route.Namespace, "name", route.Name)
+			// Extending the lease ensures a delay in future work ONLY for followers. Unlike in
+			// performIngressConditionUpdate, it's not logical to invoke findMostRecentIngress here and expect the last
+			// update to be from our router. This is because performIngressConditionRemoval *removes* a condition
+			// without providing a LastTransitionTime on a condition for us to track previous actions.
+			lease.Extend(workKey)
+			return writerlease.None, false
+		}
+
+		// If the tracker determines that another process is attempting to update the ingress to an inconsistent
+		// value, skip updating altogether and rely on the next resync to resolve conflicts. This prevents routers
+		// with different configurations from endlessly updating the route status.
+		// TRICKY: The tracker keys off of the route UID, not the workKey.
+		if tracker.IsChangeContended(contentionKey(route.UID), now, original) {
+			log.V(4).Info("skipped update due to another process altering the route with a different ingress status value", "action", action, "workKey", workKey, "original", original)
 			return writerlease.Release, false
-		case errors.IsConflict(err):
-			// just follow the normal process, and retry when we receive the update notification due to
-			// the other entity updating the route.
-			log.V(4).Info("updating route status failed due to write conflict", "action", action, "namespace", route.Namespace, "name", route.Name)
-			return writerlease.Release, true
-		default:
-			utilruntime.HandleError(fmt.Errorf("Unable to write router status for %s/%s: %v", route.Namespace, route.Name, err))
-			return writerlease.Release, true
 		}
+
+		return handleRouteStatusUpdate(context.TODO(), action, oc, route, latest, tracker)
 	})
 }
```

Comment:
> Yes I like that. I had the idea to do a callback, but got stuck following through with it, but this was I wanted.
> 
> > Maybe better left as a follow-up as it would be nontrivial to check for correctness.
> 
> Could you elaborate? It appear "correct" for our two use-cases (`performIngressConditionUpdate`, `performIngressConditionRemoval`), but are you saying we would need to explore "correctness" for other (future) use cases?
> 
> I ask because I prefer this logic, and don't exactly see how any more "non-trivial" it is than what I currently have.

I mean that it would be nontrivial for someone reviewing such a change to verify that the behavior of what you have and the behavior of what I suggested is identical.  If you're confident that my late-night refactoring is correct (or that your tests would catch any mistakes), feel free to use it.  * grin *.